### ISSUE TITLE
Is the name `HookState` good?

### ISSUE THREAD

**Issue by vensauro** [avkonst/hookstate#78]
After the "launch" of the recoil.js, i receive in a group of the telegram the link for this library.
and this lib is awesome, but after i search in the google 4 times to this, i did not find the page and the github
all results is for react vanilla useState, and articles of how use the react useState

is the name hookstate realy good? its okay that is straight to the point like react hook form, but collides with other obvious results.

feel free to close this issue if is not open for discursions

**Comment by avkonst** [comment]
Hi, maybe not. But not sure what people are searching for when what to find Hookstate lib. What would you search for? How would you name it? Maybe you can ask somebody, who sent you a link in the telegram, how he/she found it?

**Comment by avkonst** [comment]
BTW, when I search in google, duckduckgo, etc. by term "hookstate", all 10 top results point to the hookstate.js.org website. What term do you use when search? What search provider do you use?

**Comment by avkonst** [comment]
PS: hookstate is better than recoil ;)

**Comment by vensauro** [comment]
Oh, i looked now, and realy if i search `hookstate` show the website, but if i search `hook state` dont show, maybe the problem is me, but i will ask for who sended the link for me

**Comment by vensauro** [comment]
he founded searching for `state management react` and `awesome react`, realy the name is not a problem, maybe people is not searching for redux/mobx alternatives.
(i not like redux/mobx/xxx because they are much complex for the majority of cases, but i liked this because is simple and powerfull)

**Comment by avkonst** [comment]
If you like it please consider to spread few words about it to help the
project to grow. Thanks

**Comment by smooJitter** [comment]
@avkonst when you say hookstate is better than recoil, are you talking about the name, or are you talking about the approach to state management?  I am ready to make a commitment to a state management solution beyond context-api? I am having trouble managing rerenders and so far, this looks promising despite.

Specifically, I have many components tied to one context, for example, my UserContext has individual components to display each attribute on a user record, i.e., <UserFirstName/>, <UsertLastName />.  I also have input components for each attribute, for example <InputUserFirstName /> and <InputUserLastName />.  The input components manage updates to the UserContext.   

Because my UserContext contains at least 15 attributes, as you can imagine, this leads to excessive re-renders.  I believe this HookState can help with this.  Recoil also sounds promising, my concern with recoil is simple... I don't understand the concept at all. Atoms created anywhere and called from anywhere is a prelude to spaghetti code but more importantly, I don't know if it suffers from the same re-rendering issues.  Thus, when you say Hookstate is better,   It makes me wonder if it's better for my use case.

**Comment by avkonst** [comment]
> when you say hookstate is better than recoil, are you talking about the name, or are you talking about the approach to state management? 

I am talking about the approach to state management.

> Thus, when you say Hookstate is better, It makes me wonder if it's better for my use case.

It is better for your usecase.

1. I have only looked in recoil briefly to realise that it uses selector functions for rerender optimisation, which does not really go far from Redux really. Hookstate optimises rerendering without any selector functions, so much easier to use.
2. Recoil rerender optimisation, even if you use selectors, does not let you optimise parent component rerendering (ie. lifted up state) as Hookstate does. Read more what you would need to do with Redux (same with Recoil) for parent component states https://hookstate.js.org/docs/scoped-state#it-is-unique (you probably would need to read the whole page to understand the last bit better). Also see this: https://hookstate.js.org/docs/performance-intro .  And even if you do all the "workarounds" required by Redux (and I believe the same applicable for Recoil), it will not scale as well as Hookstate scales it to 10x of thousands of children components.

Hope it helps. Let me know if you have got more questions.

**Comment by ost12666** [comment]
I started with learning Redux and could not bring myself to use it, was too much. Then found MobX which was great but Hooks kind of killed it. Then I used reducx-toolkit for a production project since it is much simpler than Redux to use and great for TypeScript. Now for a POC I could not bring myself to start redux-toolkit again so I looked at Recoil. I thought I would use it but then I found you cannot change state not from a component and this was a deal breaker. Then I found hookstate that seems to easier at least a Recoil, maybe as efficient and allow me to do change state outside of a component. I wish it was more popular and have more contributors but for a POC its ok.

**Comment by avkonst** [comment]
@ost12666 please help to make it popular... write a blog? I have no idea how to make it popular. I enjoyed creating it, but I am not a sales person, you know :)

**Comment by ost12666** [comment]
I just started to use it and I liked that I can sort it out in minutes but I have much more to do and use it a lot more to know exactly how it is. I also liked the docs. I already recommended some people to take a look. The largest obstacle for production projects is popularity.

**Comment by avkonst** [comment]
Yes. People used to judge by popularity, unfortunately... That is why some
crap projects do well and some great are barely used :) there are other
important metrics in software except popularity...

**Comment by smooJitter** [comment]
@avkonst this package is absolutely amazing.  This simplicity is astonishing.   I have to admit.  the absence of boilerplate makes me nervous

**Comment by smooJitter** [comment]
@avkonst I am hearing a lot about "concurrent mode", Recoil claims to be concurrent mode ready and argues that other state management packages are not.   Do you have any idea what that means

**Comment by avkonst** [comment]
Yes, I know about it. It is about making sure that the rendered view across multiple subscribed components is consistent at any given point in time with the state, if a state is update more than once in a row. At the moment I am not convinced that "the problem" they solve is really an important problem. It was not a problem in my apps. It might be a priority for some of the applications. I am waiting for when they stabilise the api to see if Hookstate needs it. Last time I looked at it, it required snapshoting the state object (which is expensive operation, and Hookstate does not do it in contrast to Redux). Hookstate has got 2 other methods to manage rerendering, which do not exist in Redux or Recoil. The 1st is the ability to lift the state to a parent component and optimise it with help of scoped state. It gives an option to rerender the entire parent at once if a consistent view is required across multiple children (it might be actually faster than snapshoting the entire state in some cases). The second option is to do a batch update to a state, if multiple nested fields need to be updated at the same time. The last one should theoretically reduce the chances of inconsistent views, but does not guarantee it. The complete approach would be to integrate with the react api for mutable states and let the react drive snapshots. Will the solution to "the problem" be worth the performance impact? Not sure.. but maybe Hookstate will give an option to opt in to the this behaviour if a client needs it. If you need "the problem" solved, please open the separate related ticket. Thanks.
